File: ios/Audio/ObsidianAudioModule.swift
import Foundation

/// Headless Audio player. Works as an RCTBridgeModule (Paper) and as a
/// TurboModule (New Architecture) by conforming to the generated
/// `ObsidianAudioSpec` protocol when codegen is enabled.
@objc(ObsidianAudio)
class ObsidianAudio: RCTEventEmitter {
  private let engine = ObsidianAudioEngine()
  private var listeners = Set<String>()

  override init() {
    super.init()
    engine.delegate = self
  }

  @objc override static func requiresMainQueueSetup() -> Bool { false }

  override func supportedEvents() -> [String] {
    ["onState", "onProgress", "onEnded", "onError"]
  }

  // MARK: - Listener bookkeeping (TurboModule + bridge)

  @objc func addListener(_ eventName: String) { listeners.insert(eventName) }
  @objc func removeListeners(_ count: Double) {
    if Int(count) >= listeners.count { listeners.removeAll() }
  }

  private func emit(_ name: String, _ body: [String: Any]) {
    if listeners.contains(name) || listeners.isEmpty {
      sendEvent(withName: name, body: body)
    }
  }

  // MARK: - Commands

  @objc(load:) func load(_ sourceJson: String) {
    guard let data = sourceJson.data(using: .utf8),
          let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
          let uri = obj["uri"] as? String else { return }
    engine.load(uri, headers: obj["headers"] as? [String: String], drmLicenseUri: obj["drmLicenseUri"] as? String)
  }
  @objc func play() { engine.play() }
  @objc func pause() { engine.pause() }
  @objc func stop() { engine.stop() }
  @objc(seek:) func seek(_ seconds: Double) { engine.seek(seconds) }
  @objc(setRate:) func setRate(_ rate: Double) { engine.setRate(rate) }
  @objc(setVolume:) func setVolume(_ volume: Double) { engine.setVolume(volume) }
  @objc(setMuted:) func setMuted(_ muted: Bool) { engine.setMuted(muted) }
  @objc(setLoop:) func setLoop(_ loop: Bool) { engine.setLoop(loop) }

  @objc func getCurrentState(_ resolve: @escaping RCTPromiseResolveBlock,
                             reject: @escaping RCTPromiseRejectBlock) {
    resolve(engine.currentStateJson())
  }

  // MARK: - New Architecture TurboModule conformance

  #if RCT_NEW_ARCH_ENABLED
  // Conformance to the generated ObsidianAudioSpec protocol (see bottom).
  #endif
}

extension ObsidianAudio: ObsidianAudioEngineDelegate {
  func audioState(_ payload: [String: Any]) { emit("onState", ["stateJson": stateJson(payload)]) }
  func audioProgress(position: Double, duration: Double) {
    emit("onProgress", ["position": position, "duration": duration])
  }
  func audioEnded() { emit("onEnded", [:]) }
  func audioError(_ message: String) { emit("onError", ["message": message]) }

  private func stateJson(_ payload: [String: Any]) -> String {
    (try? JSONSerialization.data(withJSONObject: payload))
      .flatMap { String(data: $0, encoding: .utf8) } ?? "{}"
  }
}

#if RCT_NEW_ARCH_ENABLED
import ObsidianMediaPlayerSpec
extension ObsidianAudio: ObsidianAudioSpec {}
#endif

File: android/src/main/java/com/obsidianmediaplayer/audio/ObsidianAudioModule.kt
package com.obsidianmediaplayer.audio

import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule
import androidx.media3.common.MediaItem
import androidx.media3.common.Player
import androidx.media3.exoplayer.ExoPlayer
import com.obsidianmediaplayer.core.ExoPlayerProvider
import org.json.JSONObject

/**
 * Headless Audio player. Registered as "ObsidianAudio" for both the legacy
 * bridge (NativeModules.ObsidianAudio) and the New Architecture TurboModule.
 * Internally uses a single ExoPlayer with a disk cache.
 */
class ObsidianAudioModule(private val ctx: ReactApplicationContext) : ReactContextBaseJavaModule(ctx) {
  override fun getName() = "ObsidianAudio"

  private val player: ExoPlayer by lazy { ExoPlayerProvider.buildPlayer(ctx) }
  private var loop = false
  private val lastState = mutableMapOf<String, Any>(
    "status" to "idle", "position" to 0.0, "duration" to 0.0,
    "rate" to 1.0, "muted" to false, "volume" to 1.0,
    "buffered" to 0.0, "inBackground" to false
  )

  private val listener = object : Player.Listener {
    override fun onPlaybackStateChanged(state: Int) {
      lastState["status"] = when (state) {
        Player.STATE_READY -> if (player.playWhenReady) "playing" else "ready"
        Player.STATE_BUFFERING -> "buffering"
        Player.STATE_ENDED -> { if (loop) { player.seekTo(0); player.play(); "playing" } else "ended" }
        else -> "loading"
      }
      if (state == Player.STATE_ENDED && !loop) send("onEnded", Arguments.createMap())
      emitState()
    }
    override fun onPlayerError(error: androidx.media3.common.PlaybackException) {
      lastState["status"] = "error"; lastState["error"] = error.message ?: "unknown"
      send("onError", Arguments.createMap().apply { putString("message", error.message) })
      emitState()
    }
    override fun onIsPlayingChanged(isPlaying: Boolean) {
      lastState["status"] = if (isPlaying) "playing" else "paused"
      emitState()
    }
  }

  init { player.addListener(listener) }

  private fun send(name: String, params: WritableMap?) {
    ctx.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java).emit(name, params)
  }
  private fun emitState() {
    send("onState", Arguments.createMap().apply { putString("stateJson", JSONObject(lastState as Map<*, *>).toString()) })
  }

  // Listener book-keeping expected by NativeEventEmitter / TurboModule
  @ReactMethod fun addListener(eventName: String) {}
  @ReactMethod fun removeListeners(count: Int) {}

  @ReactMethod fun load(sourceJson: String) {
    val obj = JSONObject(sourceJson)
    val uri = obj.getString("uri")
    val headers = obj.optJSONObject("headers")?.let { jo ->
      buildMap { jo.keys().forEach { k -> put(k, jo.getString(k)) } }
    } ?: emptyMap()
    val type = obj.optString("type", null)?.takeIf { it.isNotEmpty() }
    val cacheable = obj.optBoolean("cacheable", true)
    val drmLicenseUri = obj.optString("drmLicenseUri", null)?.takeIf { it.isNotEmpty() }
    val item = MediaItem.Builder().setUri(uri).build()
    val source = ExoPlayerProvider.buildMediaSource(ctx, item, headers, type, cacheable, drmLicenseUri)
    player.setMediaSource(source); player.prepare()
    lastState["status"] = "loading"; emitState()
  }
  @ReactMethod fun play() { player.play() }
  @ReactMethod fun pause() { player.pause() }
  @ReactMethod fun stop() { player.pause(); player.seekTo(0) }
  @ReactMethod fun seek(seconds: Double) { player.seekTo((seconds * 1000).toLong()) }
  @ReactMethod fun setRate(rate: Double) {
    lastState["rate"] = rate
    player.setPlaybackParameters(androidx.media3.common.PlaybackParameters(rate.toFloat()))
  }
  @ReactMethod fun setVolume(volume: Double) { lastState["volume"] = volume; player.volume = volume.toFloat() }
  @ReactMethod fun setMuted(muted: Boolean) { lastState["muted"] = muted; player.volume = if (muted) 0f else (lastState["volume"] as Double).toFloat() }
  @ReactMethod fun setLoop(loop: Boolean) { this.loop = loop; player.repeatMode = if (loop) Player.REPEAT_MODE_ONE else Player.REPEAT_MODE_OFF }

  @ReactMethod fun getCurrentState(p: Promise) { p.resolve(JSONObject(lastState as Map<*, *>).toString()) }

  override fun invalidate() { player.release(); super.invalidate() }
}
